# Exercises for pricing derivatives on a Binomial Tree FINA60211(A)

Pour la version française de ce fichier, [cliquez ici](bino_exercises_fr.ipynb).

This interactive notebook and the associated `pedagogical_binomial_model` python module provide you with tools to generate your own exercises in pricing options on a binomial tree. You can use this notebook to:

- learn to construct forward binomial trees
- learn to calculate probabilities of states on a binomial tree
- learn to price options by recursive single-period risk-neutral pricing and discounting
- learn to price European options by calculating risk-neutral expectations of the terminal value

The module `pedagogical_binomial_model` follows the conventions and notation of our course slides.

## Working with the module

In order to learn how to use the module, you can simply read the documentation in the `pedagogical_binomial_model.py` file or try the following `help` call:

In [17]:
import pedagogical_binomial_model as pbm
help(pbm.binomial_tree)

Help on class binomial_tree in module pedagogical_binomial_model:

class binomial_tree(builtins.object)
 |  binomial_tree(Stock_0, N_steps, h_time_step_length, r_int_rate, d_div_rate, s_volatility)
 |  
 |  Forward Binomial Tree for pricing Equity Derivatives.
 |  
 |  All math and notation follows the course slides for FINA60211(A) Princpes d'evaluation des produits derives
 |  
 |  Attributes
 |  ----------
 |  Stock_0 : float
 |      initial stock price
 |  N_steps : int
 |      number of steps on the tree model
 |  h_time_step_length : float
 |      length of a single time step (in years) on the tree: the tree describes stock price behavior over the period of length N_steps * h_time_step_length
 |  r_int_rate : float
 |      risk-free interest rate (NOT in percentage pts, i.e., for a rate of 4.2% input 0.042)
 |  d_div_rate : float
 |      dividend yield, (NOT in percentage pts, i.e., for a yield of 4.2% input 0.042)
 |  s_volatility : float
 |      annualized volatility of the sto

## Examples

### Create a binomial tree for the stock price

Create a tree to describe the evolution of the stock price:

- for three months
- in one-month steps

Use the following parameters for the tree:

- $S_0$ = 100
- $r_f$ = 2%
- $\delta$ = 1%
- $\sigma$ = 20%

In [18]:
stock_price = 100
number_of_model_steps = 3 # three months
length_of_model_step = 1/12 # a month is 1/12 years
risk_free_rate = 0.02
dividend_yield = 0.01
stock_volatility = 0.2 # annualized

stock_price_tree = pbm.binomial_tree(stock_price, number_of_model_steps, length_of_model_step, risk_free_rate, dividend_yield, stock_volatility)

To inspect the tree you can use the associated `print` method. The periods of the tree are indexed from $0$ to `number_of_model_steps`. Let's have a look at the output at time $1$.

Note that the periods to be printed have to be given as a `list`: for example, `[0, 1, 2]`.

In the output below you'll see that

- Periods are numbered and separated by headers
- For each period, all possible stock price states are printed
- All states are named following the convention that says how many `U` and `D` moves are required to reach a given state, for example `U1-D0` for the "up" state at period $1$ and `U0-D1` for the "down" state at period 1.

There is extra information in the output:

- The **"Up" Transition Probability**: $p^{\star}$ from our slides
- The **Probability of State** : the probability of achieving a given state via all possible trajectories
- The **Multi-Period Discount Factor**: the discount factor **from** the period in the header **to** the terminal period; $e^{-r_f (K-k) h}$ where the terminal period is $K$ and the current period is $k$.
- The **Single-Period Discount Factor**: $e^{-r_f h}$
- The "Up" and "Down" factors are $u$ and $d$.

In [19]:
stock_price_tree.print([1])

--------------------------------------------------------------------
In Tree At Period 1:
--------------------------------------------------------------------
In State U1-D0
Stock price = 106.031747
Probability of State = 0.485570
"Up" transition probability (p-star) = 0.485570
Multi-Period Discount Factor = 0.996672
Single-Period Discount Factor = 0.998335
"Up" factor = 1.060317
"Down" factor = 0.944687


In State U0-D1
Stock price = 94.468693
Probability of State = 0.514430
"Up" transition probability (p-star) = 0.485570
Multi-Period Discount Factor = 0.996672
Single-Period Discount Factor = 0.998335
"Up" factor = 1.060317
"Down" factor = 0.944687




### Exercise:

Go back to your slides and use the stock price parameters from the preceding section to build a forward binomial tree by hand. Then use this module to check your results.

### Price a European put option

You can use the model to price a European put option. The idea is to first construct the tree that describes the evolution of the stock price and then price the option *on that tree*.

This means that one of the option's key parameters, **time to maturity**, depends on the tree!

Let's price an option with the following parameters:

- Expiration in six months
- Strike price of \$95

#### Define the stock price tree

Our existing binomial tree, `stock_price_tree` describes the evolution of the price over three months. Therefore, we **cannot** use it to price the option, and we have to create a new tree.

We will stick to a three-period tree. This means that we need two-month periods, i.e., $h = 2/12$ on the tree.

In [20]:
stock_price = 100
number_of_model_steps = 3 # three steps
length_of_model_step = 2/12 # a month is 1/12 years
risk_free_rate = 0.02
dividend_yield = 0.01
stock_volatility = 0.2 # annualized

stock_price_tree = pbm.binomial_tree(stock_price, number_of_model_steps, length_of_model_step, risk_free_rate, dividend_yield, stock_volatility)

#### Define the option

To define a European put option, we will use the `european_put` class from the `pbm` module. Here's the call to `help(pbm.european_put)`.

An `european_put` is a `derivative`, so if you're interested, you might want to follow up by reading `help(pbm.derivative)` and the code.

In [21]:
help(pbm.european_put)

Help on class european_put in module pedagogical_binomial_model:

class european_put(derivative)
 |  european_put(strike_price)
 |  
 |  European put
 |  
 |  Method resolution order:
 |      european_put
 |      derivative
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __init__(self, strike_price)
 |      Define an European Put
 |      
 |      Parameters
 |      ----------
 |      strike_price : float
 |          strike price of the option
 |  
 |  ----------------------------------------------------------------------
 |  Methods inherited from derivative:
 |  
 |  print(self, step_indexes=None)
 |  
 |  set_tree(self, binomial_tree)
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from derivative:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)



The above help listing tells us how to define an `european_put`:
```
|  european_put(strike_price)
...
|  Methods defined here:
|  
|  __init__(self, strike_price)
|      Define an European Put
|      
|      Parameters
|      ----------
|      strike_price : float
|          strike price of the option
```

Simply, set a `strike_price` and write `my_option = pbm.european_put(strike_price)`.

In [22]:
strike_price = 95
my_euro_put = pbm.european_put(strike_price)

#### Price the option

Having defined the stock price tree and the option, we can price the option.

We do it by using the `pricing()` method of the `stock_price_tree` on the option.

In [23]:
my_euro_put = stock_price_tree.pricing(my_euro_put)

Seemingly, nothing happened.

However, `my_euro_put` now contains a full binomial tree with all incremental pricing information!

The `derivative` class and all related classes (i.e., European and American calls and puts in `pbm`) have a `print` method which is identical to the `print` method for the binomial tree that we examined above.

To learn about the put's value at time $0$, you can simply type `my_euro_put.print([0])` and see in the field `"Derivative"` that the option is worth $3.1966.

In [24]:
my_euro_put.print([0])

--------------------------------------------------------------------
In Tree At Period 0:
--------------------------------------------------------------------
In State 0
Stock price = 100.000000
Probability of State = 1.000000
"Up" transition probability (p-star) = 0.479599
Multi-Period Discount Factor = 0.990050
Single-Period Discount Factor = 0.996672
"Up" factor = 1.086886
"Down" factor = 0.923132
Derivative = 3.196583
Repl. Delta = -0.300772
Repl. Bond = 33.273770




#### Exercise: Price the option using the expected terminal payoff

For European derivatives we can use the alternative approach to pricing. With $K$ periods of length $h$ years, the price of a derivative $G(S_K)$ is:
$$
    P = e^{-r_f \times h \times K}E^{\star}\left[ G(S_K) \right]\, ,
$$
the risk-neutral expectation of the payoff, discounted at the risk-free rate.

To see the probabilities of the terminal states together with the stock prices in these states, type `my_euro_put.print([3])`.

In [25]:
my_euro_put.print([3])

--------------------------------------------------------------------
In Tree At Period 3:
--------------------------------------------------------------------
In State U3-D0
Stock price = 128.395990
Probability of State = 0.110315
"Up" transition probability (p-star) = 0.479599
Multi-Period Discount Factor = 1.000000
Single-Period Discount Factor = 0.996672
"Up" factor = 1.086886
"Down" factor = 0.923132
Derivative = 0.000000


In State U2-D1
Stock price = 109.051456
Probability of State = 0.359100
"Up" transition probability (p-star) = 0.479599
Multi-Period Discount Factor = 1.000000
Single-Period Discount Factor = 0.996672
"Up" factor = 1.086886
"Down" factor = 0.923132
Derivative = 0.000000


In State U1-D2
Stock price = 92.621429
Probability of State = 0.389651
"Up" transition probability (p-star) = 0.479599
Multi-Period Discount Factor = 1.000000
Single-Period Discount Factor = 0.996672
"Up" factor = 1.086886
"Down" factor = 0.923132
Derivative = 2.378571


In State U0-D3
Stock pr

In the `"Derivative"` field you can see the values of the payoff $G(S_K)$ which here is $\max(95 - S_K, 0)$. Then use the `Probability of State` and appropriate `Multi-Period Discount Factor` from period  $0$ to period $K$ to calculate the price of the European Put with pen and paper.

#### Extracurricular (for the nerds)

You can recover all the information that is printed on the tree to use it as data in a computer program. The following code snippets extract the risk-neutral probabilities of the states at the final node of the tree, and the corresponding payoffs.

In the end, the `.trunk` attribute of the option's `pricing_tree` is a `dictionary`.

In [26]:
type(my_euro_put.pricing_tree.trunk)

dict

In [27]:
import numpy as np
rn_probs = [my_euro_put.pricing_tree.trunk["Period 3"][state]["Probability of State"] for state in my_euro_put.pricing_tree.trunk["Period 3"].keys()]
rn_probs = np.array(rn_probs)

print("These are the risk-neutral probabilities converted to a numpy array")
print(rn_probs)

These are the risk-neutral probabilities converted to a numpy array
[0.110315   0.35910036 0.38965103 0.14093361]


In [28]:
payoffs = [my_euro_put.pricing_tree.trunk["Period 3"][state]["Derivative"] for state in my_euro_put.pricing_tree.trunk["Period 3"].keys()]
payoffs = np.array(payoffs)

print("These are the put's payoffs")
print(payoffs)

These are the put's payoffs
[ 0.          0.          2.37857115 16.33319997]


Now you can calculate the risk-neutral expectation and discount it to period $0$:

In [29]:
put_price = np.sum(rn_probs * payoffs)
put_price = my_euro_put.pricing_tree.trunk["Period 0"]["State 0"]["Multi-Period Discount Factor"] * put_price

print("The price of the put option is ${0:1.4f}.".format(put_price))

The price of the put option is $3.1966.


### Price an American put option and compare with the European option

The American option is more *optional* than the European option, because it can be exercised any time before expiration. If it's more optional... then its value should be **higher** than the European option's.

The difference between the prices of otherwise identical American and European options is called the **early exercise premium**.

We can use our binomial model to price the American option estimate the premium.

An american put option can be instantiated from the class `american_put`. We will use the same tree and strike price as before.

In [30]:
my_amer_put = pbm.american_put(95)
my_amer_put = stock_price_tree.pricing(my_amer_put)

In [31]:
my_amer_put.print([0])

--------------------------------------------------------------------
In Tree At Period 0:
--------------------------------------------------------------------
In State 0
Stock price = 100.000000
Probability of State = 1.000000
"Up" transition probability (p-star) = 0.479599
Multi-Period Discount Factor = 0.990050
Single-Period Discount Factor = 0.996672
"Up" factor = 1.086886
"Down" factor = 0.923132
Derivative = 3.243454
Exercise Derivative = False




In [32]:
my_euro_put.print([0])

--------------------------------------------------------------------
In Tree At Period 0:
--------------------------------------------------------------------
In State 0
Stock price = 100.000000
Probability of State = 1.000000
"Up" transition probability (p-star) = 0.479599
Multi-Period Discount Factor = 0.990050
Single-Period Discount Factor = 0.996672
"Up" factor = 1.086886
"Down" factor = 0.923132
Derivative = 3.196583
Repl. Delta = -0.300772
Repl. Bond = 33.273770




The price of the American put is \$3.2435 and the price of the European put is \$3.1966.

The **early exercise premium** is equal to \$0.0469!

## Further exercises

Play around with this module. Use it to answer the following questions:

### Exercise 1

What is the early exercise premium for call options if $\delta = 0 $?

### Exercise 2

What is the early exercise premium for put options if $r = 0$?

### Exercise 3

What happens with the price of an European option if the binomial model's time grid gets *denser*?

Through this we mean that we could keep the model's horizon constant but describe it in a larger number of steps. Above, we used three steps for a six-month horizon. What if we used six (i.e., monthly steps)? What if we used 180 (i.e., daily steps)?

First, define a tree where `length_of_model_step = 1/365` and the `number_of_model_steps = 180`.

Then, write a function which evaluates the Black-Scholes option pricing formula and compare the prices from the formula to the prices from a binomial model with very many steps (say, 1 step per day).

Price a European option (call or put) with both methods.

### Exercise 4

The prices of European put and call options **with the same strike and maturity** are strongly related to each other! This relation is called the **put-call parity**. It can be stated as follows:
$$
    C - P = S e^{-\delta T} - K \times e^{-r_f T}\, ,
$$
where $C$ and $P$ are, respectively, the call and put option prices, $S$ is the stock price, $K$ is the strike of both options, $r_f$ is the risk-free rate, $\delta$ is the dividend yield, and $T$ is the maturity of both options.

Use the binomial model to:

1. Demonstrate that the put-call parity relation indeed holds,
2. Examine how the discrepancy from the relation depends on $\delta$ for American options.
